### PR TITLE
feat: Catalog templates with property list

### PR DESCRIPTION
**Pull request by nandorholozsnyak**
- Property can have description and default value
- Adding multiple property at the same time with a specific format
- Listing templates returns the list of the template properties too

Associated issue: https://github.com/jbangdev/jbang/issues/1203

### CODE REVIEW

File: src/main/java/dev/jbang/catalog/Template.java [hunk 1]
```diff
@@ -14,11 +14,14 @@
 	@SerializedName(value = "file-refs")
 	public final Map<String, String> fileRefs;
 	public final String description;
+	public final Map<String, TemplateProperty> properties;
 
-	public Template(Map<String, String> fileRefs, String description, Catalog catalog) {
+	public Template(Map<String, String> fileRefs, String description, Catalog catalog,
```
Comment (quintesse) [inline src/main/java/dev/jbang/catalog/Template.java:19]: This is not wrong of course, but could you change the order a bit? The `catalog` is meant to be the last argument in the list.

File: src/main/java/dev/jbang/catalog/Catalog.java [hunk 2]
```diff
@@ -59,7 +59,7 @@ public Catalog(String baseRef, String description, ResourceRef catalogRef, Map<S
 				new Alias(a.scriptRef, a.description, a.arguments, a.javaOptions, a.dependencies, a.repositories,
 						a.classpaths, a.properties, a.javaVersion, a.mainClass, this)));
 		templates.forEach((key, t) -> this.templates.put(key,
-				new Template(t.fileRefs, t.description, this)));
+				new Template(t.fileRefs, t.description, this, new HashMap<>())));
```
Comment (quintesse) [inline src/main/java/dev/jbang/catalog/Catalog.java:62]: Use `Collections.emptyMap()` here.

File: src/main/java/dev/jbang/catalog/CatalogUtil.java [hunk 3]
```diff
@@ -9,10 +9,7 @@
 import java.nio.file.InvalidPathException;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.util.AbstractMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
+import java.util.*;
```
Comment (quintesse) [inline src/main/java/dev/jbang/catalog/CatalogUtil.java:12]: Make sure you run `./gradlew spotlessApply` before committing, no `*` imports are allowed.
I always run `./gradlew spotlessApply clean build installDist` to make sure all my code has been cleaned up.

File: src/main/java/dev/jbang/catalog/CatalogUtil.java [hunk 4]
```diff
@@ -120,22 +117,25 @@ private static void removeAlias(Catalog catalog, String name) {
 	/**
 	 * Adds a new template to the nearest catalog
 	 *
-	 * @param name The name of the new template
+	 * @param properties
```
Comment (quintesse) [inline src/main/java/dev/jbang/catalog/CatalogUtil.java:120]: Could you add some description here?

File: src/main/java/dev/jbang/catalog/CatalogUtil.java [hunk 5]
```diff
@@ -120,22 +117,25 @@ private static void removeAlias(Catalog catalog, String name) {
...
 		Path catalogFile = Catalog.getCatalogFile(null);
-		addTemplate(catalogFile, name, fileRefs, description);
+		addTemplate(catalogFile, name, fileRefs, description, properties);
 		return catalogFile;
 	}
 
 	/**
 	 * Adds a new template to the given catalog
 	 *
+	 * @param properties
```
Comment (quintesse) [inline src/main/java/dev/jbang/catalog/CatalogUtil.java:133]: Same here

File: src/main/java/dev/jbang/cli/Template.java [hunk 6]
```diff
@@ -263,6 +313,22 @@ private static void printTemplate(PrintStream out, String catalogName, Catalog c
...
 			}
 		}
+		if (showProperties && template.properties != null) {
+			for (Map.Entry<String, TemplateProperty> entry : template.properties.entrySet()) {
+				out.print(Util.repeat(" ", indent));
+				StringBuilder propertyLineBuilder = new StringBuilder()
+																		.append(Util.repeat(" ", indent + 4))
+																		.append(ConsoleOutput.cyan(entry.getKey()))
+																		.append(" = ");
+				if (StringUtils.isNotBlank(entry.getValue().getDescription())) {
```
Comment (quintesse) [inline src/main/java/dev/jbang/cli/Template.java:323]: Given the changes I suggested for the `TemplatePropertyConverter` I'd simply check for `null` here.

File: src/main/java/dev/jbang/catalog/TemplateProperty.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package dev.jbang.catalog;
+
+import com.google.gson.annotations.SerializedName;
+
+public class TemplateProperty {
+
+	private String description;
+	@SerializedName(value = "default-value")
```
Comment (quintesse) [inline src/main/java/dev/jbang/catalog/TemplateProperty.java:8]: This is just my personal preference but I'd suggest simply using `default` instead of `default-value`.

**Review by quintesse — CHANGES_REQUESTED** [review]
Looks good! Just needs a couple of minor changes.
It also needs tests and changes to the docs :-)

**Comment by nandorholozsnyak** [comment]
Thank you for the feedback @quintesse . Will take a look on these. 
Just a quick feedback, in the Contribution file maybe some code standard and requirements should be visible, I understand I have to write the tests and make sure it is well formatted, but for the formatting issues I do remember I ran the spotless plugin before committing :D I'm just setting up the git hook for it to never happen again.
Is there any proper documentation about the code layering and structure?

**Review by quintesse — APPROVED** [review]
This would already be good enough to merge IMO.
But we need a couple of more tests, especially in `TestTemplate`, just to make sure that the information gets stored and retrieved correctly. A couple of edge cases perhaps (wrongly formatted property, things like that).

**Comment by quintesse** [comment]
> Is there any proper documentation about the code layering and structure?

Not really, I'd just look at the surrounding code when you add/change things and try to follow that style as much as possible. And if you write something completely new then we tend to just let you write it as you want, as long as it is good code :-)

**Comment by quintesse** [comment]
> This would already be good enough to merge IMO.

Once it builds and tests correctly of course 😉

**Comment by maxandersen** [comment]
> > Is there any proper documentation about the code layering and structure?
> 
> 
> 
> Not really, I'd just look at the surrounding code when you add/change things and try to follow that style as much as possible. And if you write something completely new then we tend to just let you write it as you want, as long as it is good code :-)

...and we apply code formatting automatically so the way I code is just whatever I need with whatever formatting my ide uses - Then run the spotlessapply
And it will fix it. Avoids having to think about formatting :)

**Comment by quintesse** [comment]
> And it will fix it

We should do something about the indenting, though. The fact that it tries to line up with the end of the first line (or whatever it is it tries to do) often creates some pretty horrible indents 😞

**Comment by nandorholozsnyak** [comment]
I have never used Spotless before but .editorconfig. For me it was working well, of course if you are an Eclipse or Netbeans user you need a plugin for it in the IDE, but it has Gradle plugins as well. 
https://editorconfig.org/

**Comment by quintesse** [comment]
@nandorholozsnyak Oh, Spotless works well enough, it's just the chosen configuration that I'm somewhat unhappy with :-)

**Comment by nandorholozsnyak** [comment]
Hey @quintesse I added tests and small changes. I added the CLI docs as well, but I'm not sure if it would be enough or not.

**Comment by nandorholozsnyak** [comment]
Btw, locally I use Java 17 (Temurin) and the build fails with that, so this is why I was not able to spot what is not working, Spotless is not working with that, or at least some things are not okay with that.
I ran the code you mentioned, now it should be working.

**Review by quintesse — APPROVED** [review]

**Comment by quintesse** [comment]
Good job @nandorholozsnyak, thanks !

**Comment by nandorholozsnyak** [comment]
Just a quick question @quintesse , did I just screw up the Codecov reports? I see that the master is failing because of the codecov report, and I just saw your new PR and checked the code as well and saw a lot of "misses" by the codecov. Should I open a PR to fix those?

**Comment by quintesse** [comment]
Codecov is still in somewhat of an experimental phase for us right now, so don't worry too much about it. At some point we'll want to do something with that information though :-)
